### PR TITLE
The Big Restructure of Toga [WIP]

### PR DESCRIPTION
**Pull request by Ocupe**
This is PR is an attempt to restructure Toga to use a factory pattern instead of inheritance. The reason for this is to make Toga easier to test.

#### Goal
The goal is to push this PR to become the next official Toga 0.3 release. I'm going to continue the port of toga.cocoa and toga.iOS to the new form. If someone wants to jump in and help to port other platforms, your more than welcome! I'm willing to help and explain when ever needed :)

@freakboy3742

### CODE REVIEW

**Comment by Ry10p** [comment]
@Ocupe 

Hey! I am very new to contributions, and I am working on the toga-gtk widgets.

Could you please tell me how to start creating new ones for toga-gtk?

A little direction is all I'm looking for. Thanks so much in advance!

_Sent from my Samsung SM-G920V using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Ocupe** [comment]
Hey @Ry10p,
great to see that you jumped into the GTK backend :)
I'm not familiar with GTK so my advice is abstract but I hope still helpful to get you started.

I guess you want to add a widget that already exist in `toga-core` to `toga-gtk`, right?
If this is the case, you should familiarise yourself with the way other widgets are implemented in `toga-gtk`. But I guess you already did this. 

Next step would be to pick a widget, from the `cocoa` backend that you want to implement and look though the GTK docs and find a equivalent. After that you have to spend some quality time with the GTK docs to understand how you can interact with your backend (how do I instantiate the widget, what function do I have to call to set a label ...). 

Most of the widgets share a common structure so use  already implemented GTK widgets as a reference and use already implemented `toga-cocoa` widgets as blueprint to see how the interface for your new widget should look like.

If you have more questions don't hesitate to ask.

**Comment by Ry10p** [comment]
@Ocupe 
Thank you so much!
I want to do do the best job possible, so I really do thank you for your guidance.
I will let you know if I run into a road block. 

Thanks again!
-Ryan Carl

File: src/cocoa/toga_cocoa/dialogs.py [hunk 1]
```diff
@@ -75,7 +73,7 @@ def stack_trace(window, title, message, content, retry=False):
 
     if retry:
         alert.addButtonWithTitle_('Retry')
-        alert.addButtonWithTitle_('Quit')
+        alert.addButtonWithTitle_('Cancel')
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/dialogs.py]: Is this change intentional?

File: src/cocoa/toga_cocoa/dialogs.py [hunk 2]
```diff
@@ -1,11 +1,9 @@
 from .libs import *
 
-from .widgets.icon import Icon
-
 
 def info(window, title, message):
     alert = NSAlert.alloc().init()
-    alert.icon = Icon.app_icon._impl
+    alert.icon = window.app._impl._native
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/dialogs.py]: Is this reference (and subsequent ones) to `_impl._native` accurate?

File: src/cocoa/toga_cocoa/factory.py [hunk 3]
```diff
@@ -1,63 +1,56 @@
-from .app import App
-from .command import Group, Command, CommandSet
+import os
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/factory.py]: Why is `os` needed?

File: src/cocoa/toga_cocoa/font.py [hunk 4]
```diff
@@ -1,7 +1,11 @@
...
+
+class Font():
+    def __init__(self, interface):
+        self.interface = interface
+        self.interface._impl = self
+        self.create()
 
     def create(self):
-        self._impl = NSFont.fontWithName_size_(self.family, self.size)
+        self._native = NSFont.fontWithName_size_(self.interface.family, self.interface.size)
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/font.py]: Is the `_native` naming consistent?

File: src/cocoa/toga_cocoa/widgets/box.py [hunk 5]
```diff
@@ -1,18 +1,9 @@
...
-        self._create()
-
+class Box(Widget):
     def create(self):
-        # # self._impl.setWantsLayer_(True)
-        # # self._impl.setBackgroundColor_(NSColor.blueColor)
-
-        self._constraints = None
+        self.native = None
+        self.constraints = None
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/widgets/box.py]: Would it be better to store these two as properties of `cocoa.Widget`?

File: src/cocoa/toga_cocoa/widgets/button.py [hunk 6]
```diff
@@ -1,67 +1,39 @@
...
-    def _set_label(self, label):
-        self._impl.setTitle_(self.label)
+    def set_label(self, label):
+        self.native.setTitle_(label)
         self.rehint()
 
-    def _set_enabled(self, value):
-        self._impl.setEnabled_(self.enabled)
-
-    def _set_background_color(self, background_color):
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/widgets/button.py]: Where has this functionality moved to?

File: src/cocoa/toga_cocoa/widgets/splitcontainer.py [hunk 7]
```diff
@@ -14,48 +12,46 @@ def splitView_resizeSubviewsWithOldSize_(self, view, size: NSSize) -> None:
     def splitViewDidResizeSubviews_(self, notification) -> None:
         # If the window is actually visible, and the split has moved,
         # a resize of all the content panels is required.
-        if self._interface.window and self._interface.window._impl.isVisible:
-            # print("SPLIT CONTAINER LAYOUT CHILDREN", self._interface._containers[0]._impl.frame.size.width, self._interface._containers[1]._impl.frame.size.width)
-            self._interface._update_child_layout()
-            self._interface.on_resize()
+        if self.interface.interface.window and self.interface.interface.window._impl.native.isVisible:
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/widgets/splitcontainer.py]: There's some naming inconsistency here - `self.interface` is actually the *implementation*.

File: src/core/tests/widgets/test_passwordinput.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        self.value = 'Password'
+        self.placeholder = 'Placeholder'
+        self.readonly = False
+
+        self.password_input = toga.PasswordInput(factory=self.factory)
+
+    def test_factory_called(self):
+        self.factory.PasswordInput.assert_called_with(interface=self.password_input)
+
+    # def test_setting_readonly_invokes_impl_method(self):
```
Comment (freakboy3742) [inline src/core/tests/widgets/test_passwordinput.py]: Are these tests off for a reason?

File: src/core/toga/command.py [hunk 9]
```diff
@@ -40,17 +42,20 @@ def __init__(self, action, label,
 
         self._enabled = self.action is not None
 
+        self._widgets = []
+
+        self.factory = get_platform_factory()
+        self._impl = self.factory.Command(interface=self)
```
Comment (freakboy3742) [inline src/core/toga/command.py]: Is an implementation level for Command required?

File: src/core/toga/widgets/image.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+from .base import Widget
+from ..platform import get_platform_factory
+
+
+class Image(object):
+    def __init__(self, path, factory=None):
+        if factory:
+            self.factory = factory
+        else:
+            self.factory = get_platform_factory()
```
Comment (freakboy3742) [inline src/core/toga/widgets/image.py]: I think this call can be simplified.

File: src/core/toga/widgets/splitcontainer.py [hunk 11]
```diff
@@ -59,14 +59,14 @@ def content(self, content):
             widget._update_layout()
             widget.app = self.app
             widget.window = self.window
-
+            # Fixme move all container stuff to _impl level.
```
Comment (freakboy3742) [inline src/core/toga/widgets/splitcontainer.py]: Is this still a TODO?

File: src/core/toga/widgets/webview.py [hunk 12]
```diff
@@ -2,71 +2,57 @@
...
         self._url = root_url
-        self._set_content(root_url, content)
-
-    def _set_url(self, value):
-        raise NotImplementedError('Webview widget must define _set_url()')
-
-    def _set_content(self, root_url, content):
-        raise NotImplementedError('Webview widget must define _set_content()')
-
-    def evaluate(self, javascript):
```
Comment (freakboy3742) [inline src/core/toga/widgets/webview.py]: Evaluate should still be an interface level function - it's API that is publicly exposed.

File: src/cocoa/toga_cocoa/widgets/button.py (hunk 6 above)
Comment (Ocupe) [inline src/cocoa/toga_cocoa/widgets/button.py]: I moved it into cocoa.Widget. We talked about it and came to the conclusion that every Widget should have the ability to be enabled and disabled.

File: src/core/tests/widgets/test_passwordinput.py (hunk 8 above)
Comment (Ocupe) [inline src/core/tests/widgets/test_passwordinput.py]: Yes, because `PasswordInput` inherits most of its functionality from `TextInput` and because we only mock the _impl `PasswordInput` these tests would fail. I should probably remove them, right?

File: src/core/toga/command.py (hunk 9 above)
Comment (Ocupe) [inline src/core/toga/command.py]: I'm not sure. At the moment we only load the icon for a command at the impl level.

File: src/core/toga/widgets/splitcontainer.py (hunk 11 above)
Comment (Ocupe) [inline src/core/toga/widgets/splitcontainer.py]: I moved the container stuff to _impl level. But I didn't managed to set the min_width/height constraints.

**Comment by Ocupe** [comment]
@freakboy3742 I implemented all your changes.

File: docs/Makefile [hunk 13]
```diff
@@ -6,18 +6,23 @@ SPHINXOPTS    =
...
+PORT = 8001
+
 # the i18n builder cannot share the environment and doctrees with the others
 I18NSPHINXOPTS  = $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) .
 
 .PHONY: help clean html dirhtml singlehtml pickle json htmlhelp qthelp devhelp epub latex latexpdf text man changes linkcheck doctest gettext
 
 help:
 	@echo "Please use \`make <target>' where <target> is one of"
+	@echo "  instal     to create a virtualenv called 'env' in the docs directory"
```
Comment (tbodt) [inline docs/Makefile]: typo

**Comment by freakboy3742** [comment]
@Ocupe I've just turned on Beekeeper for Toga, so all the tests will now be executed whenever you update this PR. 

It looks like we're currently getting failures on the tests for toga/core because Icons are getting created on the wrong platform; this will be masked if you're testing locally because you'll probably have toga_cocoa installed when running the test suite. The test machines are Linux, so they're failing due to toga_gtk being missing.

**Comment by Ocupe** [comment]
@freakboy3742 First of all – nice!! Second, I'm back from my holidays and I'm ready to finally get this huge mountain of code finalized to be merged.

**Comment by freakboy3742** [comment]
Ok. This malarkey has gone on long enough. Time to MERGE!! :-)
